---
  title: "The biocMultiAssay Project"
  author: |
    | The Waldron Lab
    |  Hunter College
    |   CUNY School of Public Health
  date: "`r format(Sys.time(), '%B %d, %Y')`"
  fontsize: 10pt
  output: 
    beamer_presentation:
      theme: "CambridgeUS"
      colortheme: "dolphin"
      fonttheme: "structureitalicserif"
      includes: 
        in_header: ~/git/biocMultiAssay/vignettes/headers/hunter-header.txt
---

```{r, echo=FALSE, message=FALSE, results='hide'}
suppressPackageStartupMessages(library(biocMultiAssay))
suppressPackageStartupMessages(library(Biobase))
suppressPackageStartupMessages(library(SummarizedExperiment))
masPheno <- data.frame(sex=c("M", "F", "M", "F"),
						  age=38:41,
						  row.names=c("Jack", "Jill", "Bob", "Barbara"))
arraydat <- matrix(seq(101, 108), ncol=4, dimnames=list(c("ENST00000294241", "ENST00000355076"), c("array1", "array2", "array3", "array4")))
arraypdat <- as(data.frame(slope53=rnorm(4), row.names=c("array1", "array2", "array3", "array4")), "AnnotatedDataFrame")
exprdat <- ExpressionSet(assayData=arraydat, phenoData=arraypdat)
## ------------------------------------------------------------------------
exprmap <- data.frame(master=rownames(masPheno)[c(1, 2, 4, 3)],
					   assay=c("array1", "array2", "array3", "array4"), stringsAsFactors = FALSE)

## ------------------------------------------------------------------------
methyldat <- matrix(1:10, ncol=5, 
                     dimnames=list(c("ENST00000355076", "ENST00000383706"),
                                   c("methyl1", "methyl2", "methyl3", "methyl4", "methyl5")))

## ------------------------------------------------------------------------
methylmap <- data.frame(master = c("Jack", "Jack", "Jill", "Barbara", "Bob"),
                        assay = c("methyl1", "methyl2", "methyl3", "methyl4", "methyl5"), stringsAsFactors = FALSE)

## ------------------------------------------------------------------------
microdat <- matrix(201:212, ncol=3, 
                    dimnames=list(c("hsa-miR-21", "hsa-miR-191", "hsa-miR-148a", "hsa-miR148b"), 
                                  c("micro1", "micro2", "micro3")))

## ------------------------------------------------------------------------
micromap <- data.frame(master = c("Jack", "Barbara", "Bob"),
                        assay = c("micro1", "micro2", "micro3"), stringsAsFactors = FALSE)

## ------------------------------------------------------------------------
suppressPackageStartupMessages(library(GenomicRanges))
gr1 <-
  GRanges(seqnames = "chr3", ranges = IRanges(58000000, 59502360), #completely encompasses ENST00000355076
          strand = "+", score = 5L, GC = 0.45)
gr2 <-
  GRanges(seqnames = c("chr3", "chr3"),
          ranges = IRanges(c(58493000, 3), width=9000), #first is within ENST0000035076
          strand = c("+", "-"), score = 3:4, GC = c(0.3, 0.5))
gr3 <-
  GRanges(seqnames = c("chr1", "chr2"),
          ranges = IRanges(c(1, 4), c(3, 9)),
          strand = c("-", "-"), score = c(6L, 2L), GC = c(0.4, 0.1))
grl <- GRangesList("gr1" = gr1, "gr2" = gr2, "gr3" = gr3)
names(grl) <- c("snparray1", "snparray2", "snparray3")

## ------------------------------------------------------------------------
rangemap <- data.frame(master = c("Jack", "Jill", "Jill"), 
						assay = c("snparray1", "snparray2", "snparray3"), stringsAsFactors = FALSE)

## ------------------------------------------------------------------------
suppressPackageStartupMessages(library(SummarizedExperiment))
nrows <- 5; ncols <- 4
counts <- matrix(runif(nrows * ncols, 1, 1e4), nrows)
rowRanges <- GRanges(rep(c("chr1", "chr2"), c(2, nrows - 2)),
                     IRanges(floor(runif(nrows, 1e5, 1e6)), width=100),
                     strand=sample(c("+", "-"), nrows, TRUE),
                     feature_id=sprintf("ID\\%03d", 1:nrows))
names(rowRanges) <- letters[1:5]
colData <- DataFrame(Treatment=rep(c("ChIP", "Input"), 2),
                     row.names= c("mysnparray1", "mysnparray2", "mysnparray3", "mysnparray4"))
rse <- SummarizedExperiment(assays=SimpleList(counts=counts),
                            rowRanges=rowRanges, colData=colData)

## ------------------------------------------------------------------------
rangemap2 <- data.frame(master = c("Jack", "Jill", "Bob", "Barbara"), 
                        assay = c("mysnparray1", "mysnparray2", "mysnparray3", "mysnparray4"), stringsAsFactors = FALSE)

## ------------------------------------------------------------------------
listmap <- list(exprmap, methylmap, micromap, rangemap, rangemap2)
names(listmap) <- c("Affy", "Methyl 450k", "Mirna", "CNV gistic", "CNV gistic2")
listmap

## ---- message = FALSE----------------------------------------------------
dfmap <- biocMultiAssay:::.convertList(listmap)
toListMap(dfmap, "assayname")

## ------------------------------------------------------------------------
objlist <- list("Affy" = exprdat, "Methyl 450k" = methyldat, "Mirna" = microdat, "CNV gistic" = grl, "CNV gistic2" = rse)

## -----------------------------------------------------------
myMultiAssay <- MultiAssayExperiment(objlist, masPheno, dfmap)
```

biocMultiAssay Bioconductor package
=================================================

- Objectives
- Data types, classes, methods
- Planned development


biocMultiAssay objectives
=================================================

- Interrogation of several molecule types in one set of biological samples
- Build on existing methods for storing data types
    + harmonize methods of relevant Bioconductor classes
- Support "ragged" datasets
    + partially overlapping features and samples
    + replicated features and samples
    + ID-based and genomic range-based data, with metadata
- Simplify complicated bookkeeping
    + manipulate multiple assays + metadata simultaneously
- Plan for on-disk and remote representations of big data

biocMultiAssay design
=================================================

Primary design elements:

1. `MultiAssayExperiment`: the main container
2. `elist`: experimental data
3. `stage`: staging for subsetting operations

`MultiAssayExperiment` class structure
=================================================

"Under the hood" of a `MultiAssayExperiment`:

| Slot        | Class      | What is it?                     |
|-------------|------------|---------------------------------|
| elist       | elist      | Assay data and metadata         |
| masterPheno | DataFrame  | Specimen-level phenotype data   |
| sampleMap   | DataFrame  | Mapping specimens to assays     |
| metadata    | ANY        | Whole-experiment level metadata |
| drop        | list       | Data manimulation provenance    |

`MultiAssayExperiment` constructor
=================================================

A minimal "ragged" example from two ExpressionSets:

```{r}
miniMA <- MultiAssayExperiment(list(eset1=exprdat[1, 1:3], 
                                    eset2=exprdat[1:2, 3:4]))
```

A fuller example with specimen data `masPheno` and a `dfmap` relating different sample naming schemes:
```{r}
myMultiAssay <- MultiAssayExperiment(objlist, masPheno, dfmap)
```

`elist` slot, class, and verb
=================================================

* Experimental data are stored as an `elist`
* Inherits from S4Vectors::SimpleList
    + adds a `show` method
* Generic `elist()` function extracts or creates `elist` object

* Validity checks: 
    + require `features()`, `samples()`, `subsetFeatures()`, `subsetSamples()` methods
    + consistency checks on `samples()` output when part of a `MultiAssayExperiment`

`elist` slot, class, and verb
=================================================

`elist` the constructor:
```{r, eval=FALSE}
elist(list(affy=myExpressionSet, muts=mySE, cna=myGRL))
```

`elist the extractor`:
```{r}
elist(myMultiAssay)
```

`masterPheno` slot and verb
=================================================

```{r}
masterPheno(myMultiAssay)
```

* Validity checks: must be a DataFrame
    + rownames are sample IDs

`sampleMap` slot and verb
=================================================

* sampleMap relates biological specimens to assays
    + 3-column DataFrame: 
\tiny    
```{r}
head(sampleMap(myMultiAssay))
```

* Validity checks: 
    + items in `master` column must be found in rownames(masterPheno)
    + items in `assay` column must be found in samples(elist[[i]])
    + `assayname` column must contain only and all `elist` element names

`stage` class and verb
=================================================

- Set up a subset operation by identifiers
- Staging types include `samples`, `features`, and `assays`
- Feature staging can be ID-based or range-based
- A two-step subset procedure:
    + provides a way to log changes to the `MultiAssayExperiment` object
    + is compatible with on-disk and remote "big data"

`stage` class and verb
=================================================

* `stage` class structure:

| Slot  | Class     |
|-------|-----------|
| query | ANY       |
| keeps | list      |
| drops | list      |
| type  | character |

`stage` class and verb
=================================================

Staging assays:
```{r}
stage(myMultiAssay, "Affy", "assays")
```

`stage` class and verb
=================================================

Staging samples:
\small
```{r}
stage(myMultiAssay, c("Jack", "Jill"), "samples")
```

`stage` class and verb
=================================================

Staging features by ID:

```{r}
stage(myMultiAssay, "ENST00000355076", "features")
```
`stage` class and verb
=================================================

Staging features by ranges:

```{r}
rangeSubset <- GRanges(seqnames = c("chr1"), 
        strand = c("-", "+", "-"), 
        ranges = IRanges(start = c(1, 4, 6), width = 3))
stage(myMultiAssay, rangeSubset, "features",
      maxgap = 2L, type = "within")
```

`drops` slot
=================================================

* uses `stage` information to maintain complete data provenance
    + subsetting by feature, sample, and assay
    + tracks "keeps" and "drops" (using IDs)


biocMultiAssay harmonizes Bioconductor vocabulary
=================================================
- Gene expression, methylation, mutations, microRNA, proteins, etc.
- Currently supported classes include:
    + `GRangesList`, `RangedSummarizedExperiment`, `ExpressionSet`, `matrix`

```{r, eval = FALSE}
# ExpressionSet
phenoData(ExpressionSet); pData(ExpressionSet);
    exprs(ExpressionSet); ExpressionSet$variable; 
    featureNames(ExpressionSet); sampleNames(ExpressionSet)
# GRanges
seqnames(GRanges); strand(GRanges); ranges(GRanges); 
    mcols(GRanges); seqinfo(GRanges)
# RangedSummarizedExperiment
RSE <- RangedSummarizedExperiment; rowRanges(RSE); 
  colData(RSE); mcols(RSE); granges(RSE); strand(RSE);
  ranges(RSE); assays(RSE)
```


biocMultiAssay harmonizes Bioconductor vocabulary
=================================================

```{r}
library(biocMultiAssay)
methods("features")
```

__Note:__ also provides **`samples()`**, **"`[`"** (bracket), **`subsetSample()`**, and **`subsetFeature()`** functions to harmonize classes

Upcoming priorities
=================================================

- Support for on-disk range-based data, _e.g._ VcfStack (Vince Carey)
- Support for on-disk ID-based data, _e.g._ SQL backend
- Subsetting by networks and gene sets
    + _e.g._ protein interaction networks, miRNA regulation, RNA-binding proteins... (directed networks)
    + Gene Ontology categories
- Subsetting across ID-based and range-based datasets
- automated mapping of gene identifiers

Conclusions
=================================================

- Harmonizes subsetting and matching across different data types
- Links features and samples to clinical / specimen identifiers
- In the process of curating TCGA and other publicly available data:
    + Curated Cancer Subtypes
    + Full clinical data
    + Pathological imaging information

- http://github.com/vjcitn/biocMultiAssay
- dynamic API at:
```{r, eval=FALSE}
biocMultiAssay::API()
```

Thank you!
=================================================

1. Others in my lab:
    + Marcel Ramos, MPH (primary programmer)
    + Stephie Louis, Rimsha Azar, Ahou Eby (undergraduate interns, TCGA subtype curation)
    + Tiffany Chan (MPH candidate, TCGA clinical data curation)

2. Project collaborators:
    + Martin Morgan (PI)
    + Vincent Carey
    + Kasper Hansen